### PR TITLE
Image Statistics Module and Histograms

### PR DESCRIPTION
**Pull request by AaronViviano**
Hello,
I've added a new Image Statistics Module for functions and stucts related to information about images. Currently this is limited to generating a histogram struct, but eventually functionality could be added for [Histogram Equalization](https://en.wikipedia.org/wiki/Histogram_equalization) and [Thresholding](https://en.wikipedia.org/wiki/Thresholding_%28image_processing%29).

This is my first contribution to an open source project, so feel free to let me know if I need to do anything else to correctly create this pull request.

Thank you,
Aaron

### CODE REVIEW

**Comment by nwin** [comment]
Thank you for your contribution. As it is now, the utility of that particular histogram struct is unclear to me. I wouldn’t feel comfortable adding it without seeing it applied in a concrete image operation like the thresholding you mentioned.

File: src/imagestat/mod.rs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+/// Supports both color and greyscale images.
+pub struct Histogram {
+	/// The color channels for the histogram.
+	pub channels: Vec<HistogramChannel>,
+}
+
+impl Histogram {
+	/// Sets the count for a particular value of pixels.
+    pub fn set_count(&mut self, channel: usize, value: usize, count: i32)
+	{
```
Comment (nwin) [inline src/imagestat/mod.rs]: This bracket should be on the previous line. You’re also mixing tabs with spaces, please use 4 spaces everywhere.

File: src/imagestat/mod.rs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	
+	// Setup the histogram with the correct number of channels.
+	for x in 0..number_of_channels {
+		h.channels.push(HistogramChannel { values: std::iter::repeat(0).take(num::pow(2, bits)).collect::<Vec<_>>() });
+	}
+	
+	let (width, height) = im.dimensions();
+	
+	// Go through the image and record the pixel data.
+	// Todo: Is there a way to simply iterate through pixels using a Dynamic Image?
```
Comment (nwin) [inline src/imagestat/mod.rs]: Yes, you can just use `im.pixels()`

File: src/imagestat/mod.rs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+/// Create a histogram from an image.
+pub fn create_histogram(im: &DynamicImage) -> Histogram {
+	let color = im.color();
+	let mut h = Histogram {channels: Vec::new()};
+	let bits = color::bits_per_pixel(color);
+	let number_of_channels = color::num_components(color);
+	
+	// Setup the histogram with the correct number of channels.
+	for x in 0..number_of_channels {
+		h.channels.push(HistogramChannel { values: std::iter::repeat(0).take(num::pow(2, bits)).collect::<Vec<_>>() });
```
Comment (nwin) [inline src/imagestat/mod.rs]: You can just use `vec![0; 2 << bits]` to create that vector.

File: src/imagestat/mod.rs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+use color;
+use color::ColorType;
+use color::Rgba;
+use dynimage::DynamicImage;
+use image::GenericImage;
+use num;
+use std;
+
+#[allow(missing_copy_implementations)]
```
Comment (nwin) [inline src/imagestat/mod.rs]: That attribute is unnecessary.

**Comment by AaronViviano** [comment]
Your welcome. I've update the code based on your comments. I understand the desire for a concrete feature to go along with the Histogram code. My original thought was to provide an implementation for others to build off of if they needed a histogram. However, I'll see about implementing [Otsu's method](https://en.wikipedia.org/wiki/Otsu%27s_method) for automatic thresholding along with a generic thresholding function.

Should the threshold functions be in the image ops module or would you be fine with them in the image statistics module? Also if the functions are in the image ops module would you prefer a new threshold submodule to contain them?

Thank you.

**Comment by bvssvni** [comment]
This might be a better fit for the [imageproc](https://github.com/PistonDevelopers/imageproc) library.

**Comment by bvssvni** [comment]
Closing due to inactivity. I recommend taking a look at the [imageproc](https://github.com/PistonDevelopers/imageproc) library and see if it is a better fit.

**Comment by AaronViviano** [comment]
Hello Sven,
My apologies for the delay in responding. I'll take a look at the imageproc library. Its likely that the library will be a better fit for the thresholding code. Thank you for pointing it out.

Aaron

**Comment by bvssvni** [comment]
@AaronViviano I am sure @theotherphil would like some contributions!

**Pull request closed without merging**